File: reduce-array-size-to-the-half/reduce-array-size-to-the-half.py
class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        frequencies = Counter(arr)
        freq, total = [], 0
        for key, val in frequencies.items():
            freq.append(val)
            total += val
        
        freq.sort()
        ans, n = 0, len(arr)
        
        while total > n // 2:
            total -= freq.pop()
            ans += 1
        
        return ans
        

File: reduce-array-size-to-the-half/reduce-array-size-to-the-half.cpp
class Solution {
public:
    int minSetSize(vector<int>& arr) {
        unordered_map<int, int> frequencies;
        for (int i : arr) ++frequencies[i];
        priority_queue<int, vector<int>> freq;
        int total = 0, ans = 0, n = arr.size();
        for (auto& [key, val] : frequencies) {
            freq.push(val);
            total += val;
        }
        
        while (total > n / 2){
            total -= freq.top();
            freq.pop();
            ++ans;
        }
        
        return ans;
    }
};